src/vsr/client.zig

```diff
@@ -151,37 +154,8 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
             while (self.request_queue.pop()) |inflight| {
                 self.unref(inflight.message);
             }
-            assert(self.messages_available == constants.client_request_queue_max);
             self.message_bus.deinit(allocator);
-        }
-
-        pub fn on_message(message_bus: *MessageBus, message: *Message) void {
-            const self = @fieldParentPtr(Self, "message_bus", message_bus);
-            log.debug("{}: on_message: {}", .{ self.id, message.header });
-            if (message.header.invalid()) |reason| {
-                log.debug("{}: on_message: invalid ({s})", .{ self.id, reason });
-                return;
-            }
-            if (message.header.cluster != self.cluster) {
-                log.warn("{}: on_message: wrong cluster (cluster should be {}, not {})", .{
-                    self.id,
-                    self.cluster,
-                    message.header.cluster,
-                });
-                return;
-            }
-            switch (message.header.command) {
-                .pong_client => self.on_pong_client(message),
-                .reply => self.on_reply(message),
-                .eviction => self.on_eviction(message),
-                else => {
-                    log.warn("{}: on_message: ignoring misdirected {s} message", .{
-                        self.id,
-                        @tagName(message.header.command),
-                    });
-                    return;
-                },
-            }
```

Comment:
Assuming that it's possible for a client to receive a misdirected `.request`, it would incorrectly increment the message counter when unref`d.
We could mitigate it by checking the client and cluster id, but it's still possible to receive a request formatted like that (I can't think of a way for it to happen without a malicious agent, but it's not impossible).

I'm starting to wonder if it's worth keeping `client.messages_available`, or if we could just rely on `client.request_queue.count` to check if we are out of messages.

Assuming that we have calculated the correct maximum number of messages to never exhaust the message pool, even in odds cases such as retransmission (the message is still held by the send queue after leaving the request queue).

Reply:
> I can't think of a way for it to happen without a malicious agent, but it's not impossible

😂 Just discovered a particular case when the `recv_message` is unref'd during `deinit`, causing the message counter to be double incremented because the message header was left with the previous `command == .request`.

See the commit message for more info.